File: data-structures-dynamic-programming-pro/src/dp_problems.cpp
```cpp
#include "dp_problems.hpp"
#include <vector>
#include <string>
#include <map>
#include <algorithm>
#include <numeric> // For std::accumulate in some contexts, if needed

// --- Problem 1: Fibonacci Numbers ---

/**
 * @brief Computes the N-th Fibonacci number using a brute-force recursive approach.
 *        Demonstrates overlapping subproblems.
 * @param n The index of the Fibonacci number to compute (n >= 0).
 * @return The N-th Fibonacci number.
 * @complexity Time: O(2^n) - Exponential, due to recomputing same subproblems.
 *             Space: O(n) - For the recursion call stack depth.
 */
long long fibonacci_recursive(int n) {
    if (n <= 1) {
        return n;
    }
    return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2);
}

/**
 * @brief Helper function for memoized Fibonacci.
 * @param n The index of the Fibonacci number to compute.
 * @param memo Reference to the memoization table (initialized with -1 or similar).
 * @return The N-th Fibonacci number.
 */
long long fibonacci_memoized_helper(int n, std::vector<long long>& memo) {
    if (n <= 1) {
        return n;
    }
    // If already computed, return the stored result
    if (memo[n] != -1) {
        return memo[n];
    }
    // Compute and store the result
    memo[n] = fibonacci_memoized_helper(n - 1, memo) + fibonacci_memoized_helper(n - 2, memo);
    return memo[n];
}

/**
 * @brief Computes the N-th Fibonacci number using memoization (top-down DP).
 *        Avoids recomputing subproblems by storing results.
 * @param n The index of the Fibonacci number to compute (n >= 0).
 * @return The N-th Fibonacci number.
 * @complexity Time: O(n) - Each subproblem F(i) is computed once.
 *             Space: O(n) - For the memoization table and recursion call stack.
 */
long long fibonacci_memoized(int n) {
    if (n < 0) return 0; // Or throw an error for invalid input
    std::vector<long long> memo(n + 1, -1); // Initialize memo table with -1 (or other indicator)
    return fibonacci_memoized_helper(n, memo);
}

/**
 * @brief Computes the N-th Fibonacci number using tabulation (bottom-up DP).
 *        Builds the solution iteratively from base cases.
 * @param n The index of the Fibonacci number to compute (n >= 0).
 * @return The N-th Fibonacci number.
 * @complexity Time: O(n) - Single loop from 2 to n.
 *             Space: O(n) - For the DP table `dp`.
 */
long long fibonacci_tabulated(int n) {
    if (n < 0) return 0;
    if (n <= 1) return n;

    std::vector<long long> dp(n + 1);
    dp[0] = 0;
    dp[1] = 1;

    // Fill dp table iteratively
    for (int i = 2; i <= n; ++i) {
        dp[i] = dp[i - 1] + dp[i - 2];
    }
    return dp[n];
}

/**
 * @brief Computes the N-th Fibonacci number using space-optimized tabulation.
 *        Reduces space complexity to O(1) by only storing the last two values.
 * @param n The index of the Fibonacci number to compute (n >= 0).
 * @return The N-th Fibonacci number.
 * @complexity Time: O(n) - Single loop from 2 to n.
 *             Space: O(1) - Only a few variables are used.
 */
long long fibonacci_space_optimized(int n) {
    if (n < 0) return 0;
    if (n <= 1) return n;

    long long a = 0; // Represents F(i-2)
    long long b = 1; // Represents F(i-1)
    long long current_fib = 0;

    // Iterate from 2 up to n
    for (int i = 2; i <= n; ++i) {
        current_fib = a + b; // F(i) = F(i-2) + F(i-1)
        a = b;               // Update F(i-2) to F(i-1)
        b = current_fib;     // Update F(i-1) to F(i)
    }
    return current_fib;
}


// --- Problem 2: Longest Common Subsequence (LCS) ---

/**
 * @brief Computes the length of the Longest Common Subsequence (LCS) using tabulation (bottom-up DP).
 * @param text1 The first input string.
 * @param text2 The second input string.
 * @return The length of the LCS.
 * @complexity Time: O(m*n) where m is length of text1, n is length of text2.
 *             Space: O(m*n) for the 2D DP table.
 */
int longestCommonSubsequence(const std::string& text1, const std::string& text2) {
    int m = text1.length();
    int n = text2.length();

    // dp[i][j] stores the length of LCS of text1[0...i-1] and text2[0...j-1]
    std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));

    // Fill the dp table
    for (int i = 1; i <= m; ++i) {
        for (int j = 1; j <= n; ++j) {
            // If characters match, add 1 to the LCS of previous substrings
            if (text1[i - 1] == text2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
            } else {
                // If characters don't match, take the maximum LCS from
                // 1. LCS of text1[0...i-2] and text2[0...j-1] (skip char from text1)
                // 2. LCS of text1[0...i-1] and text2[0...j-2] (skip char from text2)
                dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }
    // The value at dp[m][n] is the length of LCS for full strings
    return dp[m][n];
}

/**
 * @brief Reconstructs one of the actual Longest Common Subsequence (LCS) strings.
 *        Requires filling the DP table first, then backtracking.
 * @param text1 The first input string.
 * @param text2 The second input string.
 * @return One of the LCS strings. If multiple exist, any valid one is returned.
 * @complexity Time: O(m*n) for building DP table + O(m+n) for reconstruction = O(m*n).
 *             Space: O(m*n) for the 2D DP table.
 */
std::string reconstructLCS(const std::string& text1, const std::string& text2) {
    int m = text1.length();
    int n = text2.length();

    // First, compute the DP table as in longestCommonSubsequence
    std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));
    for (int i = 1; i <= m; ++i) {
        for (int j = 1; j <= n; ++j) {
            if (text1[i - 1] == text2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
            } else {
                dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    // Now, reconstruct the LCS by backtracking through the DP table
    std::string lcs = "";
    int i = m;
    int j = n;

    while (i > 0 && j > 0) {
        // If characters match, this character is part of LCS
        if (text1[i - 1] == text2[j - 1]) {
            lcs = text1[i - 1] + lcs; // Prepend character to build LCS in correct order
            i--;
            j--;
        } else if (dp[i - 1][j] > dp[i][j - 1]) {
            // If top cell is greater, move up (character from text1 was skipped)
            i--;
        } else {
            // If left cell is greater or equal, move left (character from text2 was skipped)
            j--;
        }
    }
    return lcs;
}


// --- Problem 3: 0/1 Knapsack Problem ---

/**
 * @brief Solves the 0/1 Knapsack problem using tabulation (bottom-up DP).
 *        Finds the maximum value that can be obtained for a given capacity.
 * @param weights A vector of item weights.
 * @param values A vector of item values.
 * @param capacity The maximum capacity of the knapsack.
 * @return The maximum total value.
 * @complexity Time: O(N*W) where N is number of items, W is knapsack capacity.
 *             Space: O(N*W) for the 2D DP table.
 */
int knapsack_tabulated(const std::vector<int>& weights, const std::vector<int>& values, int capacity) {
    int n = weights.size();
    // dp[i][w] stores the maximum value that can be obtained with first i items
    // and knapsack capacity w.
    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(capacity + 1, 0));

    // Fill the dp table
    // i represents the current item being considered (from 1 to n)
    // w represents the current knapsack capacity being considered (from 0 to capacity)
    for (int i = 1; i <= n; ++i) {
        for (int w = 0; w <= capacity; ++w) {
            // Option 1: Don't include the current item (item i-1)
            // Value is the same as max value with (i-1) items and same capacity w
            dp[i][w] = dp[i - 1][w];

            // Option 2: Include the current item (item i-1), if its weight allows
            if (weights[i - 1] <= w) {
                // Value is the value of current item + max value with (i-1) items
                // and reduced capacity (w - weights[i-1])
                dp[i][w] = std::max(dp[i][w], values[i - 1] + dp[i - 1][w - weights[i - 1]]);
            }
        }
    }
    // The value at dp[n][capacity] is the maximum value for all n items and full capacity.
    return dp[n][capacity];
}

/**
 * @brief Solves the 0/1 Knapsack problem using space-optimized tabulation.
 *        Reduces space complexity to O(W) by using only two rows (or even one).
 * @param weights A vector of item weights.
 * @param values A vector of item values.
 * @param capacity The maximum capacity of the knapsack.
 * @return The maximum total value.
 * @complexity Time: O(N*W) where N is number of items, W is knapsack capacity.
 *             Space: O(W) for the 1D DP table.
 */
int knapsack_space_optimized(const std::vector<int>& weights, const std::vector<int>& values, int capacity) {
    int n = weights.size();
    // dp[w] stores the maximum value that can be obtained with items processed so far
    // and knapsack capacity w.
    // We only need the previous row's values to compute the current row.
    // By iterating 'w' from capacity down to 0, we ensure that dp[w - weights[i-1]]
    // refers to the value from the 'previous row' (i-1 items) before it's updated for current item 'i'.
    std::vector<int> dp(capacity + 1, 0);

    // Iterate through each item
    for (int i = 0; i < n; ++i) { // i is 0-indexed for weights/values
        int current_weight = weights[i];
        int current_value = values[i];

        // Iterate knapsack capacity from right to left (capacity down to current_weight)
        // This is crucial to avoid using the current item multiple times
        // and ensures we are always using values from the previous iteration (i-1 items)
        for (int w = capacity; w >= current_weight; --w) {
            // Option 1: Don't include current item (dp[w] already holds this from previous iteration)
            // Option 2: Include current item (current_value + dp[w - current_weight])
            dp[w] = std::max(dp[w], current_value + dp[w - current_weight]);
        }
    }
    return dp[capacity];
}


// --- Problem 4: Coin Change Problem (Count Ways) ---

/**
 * @brief Computes the number of ways to make change for a given amount using a set of coins.
 *        Assumes an infinite supply of each coin type (unbounded knapsack variant).
 * @param coins A vector of available coin denominations.
 * @param amount The target amount to make change for.
 * @return The number of distinct ways to make change.
 * @complexity Time: O(N*Amount) where N is number of coin denominations, Amount is target sum.
 *             Space: O(Amount) for the 1D DP table.
 */
long long coinChange_countWays_tabulated(const std::vector<int>& coins, int amount) {
    if (amount < 0) return 0; // Cannot make change for a negative amount

    // dp[i] will store the number of ways to make change for amount i
    std::vector<long long> dp(amount + 1, 0);

    // Base case: There is one way to make change for amount 0 (by taking no coins)
    dp[0] = 1;

    // Iterate through each coin
    for (int coin : coins) {
        // For each coin, iterate through amounts from the coin's value up to the target amount
        // This ensures that for each amount, we consider adding the current coin
        // The order of loops here (coins outer, amount inner) is important for counting combinations
        // If amount outer, coins inner, it would count permutations (order matters)
        for (int j = coin; j <= amount; ++j) {
            // Add the number of ways to make change for (j - coin) to dp[j]
            // This represents using the current 'coin' to contribute to amount 'j'
            dp[j] += dp[j - coin];
        }
    }
    return dp[amount];
}

```

File: coding-interview-dynamic-programming-complete/algorithms/dp_problems.py
import functools
from typing import List, Dict, Tuple, Optional

# --- Problem 1: Fibonacci Sequence ---

# Problem Description:
# Given an integer n, return the nth Fibonacci number.
# The Fibonacci sequence is defined as F(0) = 0, F(1) = 1, and F(n) = F(n-1) + F(n-2) for n > 1.

# Approach 1: Recursive with Memoization (Top-Down DP)
# Time Complexity: O(n) - Each Fibonacci number is computed once and stored.
# Space Complexity: O(n) - For the recursion stack and the memoization dictionary/cache.
@functools.lru_cache(None) # Python's built-in memoization decorator
def fibonacci_memoized(n: int) -> int:
    """
    Computes the nth Fibonacci number using recursion with memoization.
    This is a top-down dynamic programming approach.
    """
    if n <= 1:
        return n
    return fibonacci_memoized(n - 1) + fibonacci_memoized(n - 2)

# Approach 2: Iterative with Tabulation (Bottom-Up DP)
# Time Complexity: O(n) - We iterate from 2 up to n, performing constant work.
# Space Complexity: O(n) - For the DP table (array).
def fibonacci_tabulation(n: int) -> int:
    """
    Computes the nth Fibonacci number using iteration with tabulation.
    This is a bottom-up dynamic programming approach.
    """
    if n <= 1:
        return n

    # dp[i] will store the ith Fibonacci number
    dp: List[int] = [0] * (n + 1)
    dp[0] = 0
    dp[1] = 1

    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]

    return dp[n]

# Approach 3: Space-Optimized Iterative
# Time Complexity: O(n) - Same as tabulation.
# Space Complexity: O(1) - We only store the last two Fibonacci numbers.
def fibonacci_space_optimized(n: int) -> int:
    """
    Computes the nth Fibonacci number using a space-optimized iterative approach.
    This reduces the space complexity from O(n) to O(1).
    """
    if n <= 1:
        return n

    a, b = 0, 1 # a = F(i-2), b = F(i-1)

    for _ in range(2, n + 1):
        # F(i) = F(i-1) + F(i-2)
        # new_b = b + a
        # a becomes old b, b becomes new_b
        a, b = b, a + b

    return b

# --- Problem 2: 0/1 Knapsack Problem ---

# Problem Description:
# Given weights and values of N items, and a knapsack capacity W.
# Find the maximum value that can be put into the knapsack.
# Each item can either be taken or not taken (0/1 property).

# Approach 1: Recursive with Memoization (Top-Down DP)
# Time Complexity: O(N * W) - Each state (index, capacity) is computed once.
# Space Complexity: O(N * W) - For the memoization table (dictionary) and recursion stack.
def knapsack_memoized(weights: List[int], values: List[int], capacity: int) -> int:
    """
    Solves the 0/1 Knapsack problem using recursion with memoization (top-down DP).
    """
    n = len(weights)
    memo: Dict[Tuple[int, int], int] = {}

    def solve(idx: int, current_capacity: int) -> int:
        # Base Case: If no items left or no capacity, no value can be added.
        if idx == n or current_capacity == 0:
            return 0

        # Check memoization table
        if (idx, current_capacity) in memo:
            return memo[(idx, current_capacity)]

        # Option 1: Exclude the current item
        # We move to the next item, capacity remains the same.
        exclude_val = solve(idx + 1, current_capacity)

        # Option 2: Include the current item (if possible)
        include_val = 0
        if weights[idx] <= current_capacity:
            # We add its value and move to the next item with reduced capacity.
            include_val = values[idx] + solve(idx + 1, current_capacity - weights[idx])

        # Store and return the maximum of the two options
        memo[(idx, current_capacity)] = max(exclude_val, include_val)
        return memo[(idx, current_capacity)]

    return solve(0, capacity)

# Approach 2: Iterative with Tabulation (Bottom-Up DP)
# Time Complexity: O(N * W) - Two nested loops.
# Space Complexity: O(N * W) - For the 2D DP table.
def knapsack_tabulation(weights: List[int], values: List[int], capacity: int) -> int:
    """
    Solves the 0/1 Knapsack problem using iteration with tabulation (bottom-up DP).
    dp[i][j] represents the maximum value that can be obtained with items from 0 to i-1
    and a knapsack capacity of j.
    """
    n = len(weights)
    # dp table with (n+1) rows and (capacity+1) columns
    # dp[i][j] will store the max value using first `i` items with capacity `j`
    dp: List[List[int]] = [[0] * (capacity + 1) for _ in range(n + 1)]

    # Fill the DP table
    # i represents the current item being considered (from 1 to n)
    # j represents the current capacity (from 0 to capacity)
    for i in range(1, n + 1):
        for j in range(capacity + 1):
            # The current item's weight and value (using 0-based indexing for arrays)
            current_weight = weights[i - 1]
            current_value = values[i - 1]

            # Option 1: Exclude current item
            # The value is the same as the value without this item (i.e., from row i-1)
            dp[i][j] = dp[i - 1][j]

            # Option 2: Include current item (if its weight doesn't exceed current capacity)
            if current_weight <= j:
                # Value if included = current_value + max value from previous items
                # with remaining capacity (j - current_weight)
                include_val = current_value + dp[i - 1][j - current_weight]
                dp[i][j] = max(dp[i][j], include_val)

    return dp[n][capacity]

# Space-optimized version for knapsack (not explicitly requested as a separate problem,
# but a common optimization for 0/1 Knapsack)
# Time Complexity: O(N * W)
# Space Complexity: O(W) - Uses only two rows or one row (if iterating carefully)
def knapsack_space_optimized(weights: List[int], values: List[int], capacity: int) -> int:
    """
    Solves the 0/1 Knapsack problem using space-optimized iteration (bottom-up DP).
    Reduces space complexity from O(N*W) to O(W).
    """
    n = len(weights)
    # dp[j] will store the maximum value for a knapsack of capacity j.
    # We only need the previous row's values, so we can optimize to 1D array.
    dp: List[int] = [0] * (capacity + 1)

    for i in range(n):  # Iterate through each item
        current_weight = weights[i]
        current_value = values[i]

        # Iterate backwards through capacity. This is CRUCIAL for 0/1 Knapsack.
        # If we iterate forwards, an item might be considered multiple times,
        # making it an unbounded knapsack. Iterating backwards ensures that
        # when we access dp[j - current_weight], it refers to the state from the *previous* item iteration.
        for j in range(capacity, current_weight - 1, -1):
            # Option 1: Exclude the current item (dp[j] already holds this from the previous iteration)
            # Option 2: Include the current item
            # Compare with value obtained by including the current item:
            # current_value + max value from previous items with remaining capacity
            dp[j] = max(dp[j], current_value + dp[j - current_weight])

    return dp[capacity]


# --- Problem 3: Longest Common Subsequence (LCS) ---

# Problem Description:
# Given two strings `text1` and `text2`, return the length of their longest common subsequence.
# If there is no common subsequence, return 0.
# A subsequence of a string is a new string generated from the original string with some characters
# (can be none) deleted without changing the relative order of the remaining characters.
# For example, "ace" is a subsequence of "abcde".

# Approach 1: Recursive with Memoization (Top-Down DP)
# Time Complexity: O(m * n) - Each state (i, j) is computed once.
# Space Complexity: O(m * n) - For the memoization table (dictionary) and recursion stack.
def lcs_memoized(text1: str, text2: str) -> int:
    """
    Computes the length of the Longest Common Subsequence (LCS)
    using recursion with memoization (top-down DP).
    """
    len1, len2 = len(text1), len(text2)
    memo: Dict[Tuple[int, int], int] = {}

    def solve(i: int, j: int) -> int:
        # Base Case: If either string is exhausted, no more common characters.
        if i == len1 or j == len2:
            return 0

        # Check memoization table
        if (i, j) in memo:
            return memo[(i, j)]

        if text1[i] == text2[j]:
            # If characters match, they are part of the LCS. Add 1 and recurse for remaining strings.
            memo[(i, j)] = 1 + solve(i + 1, j + 1)
        else:
            # If characters don't match, we have two options:
            # 1. Skip character from text1 and try to match text2[j] with text1[i+1...].
            # 2. Skip character from text2 and try to match text1[i] with text2[j+1...].
            # Take the maximum of these two options.
            memo[(i, j)] = max(solve(i + 1, j), solve(i, j + 1))

        return memo[(i, j)]

    return solve(0, 0)

# Approach 2: Iterative with Tabulation (Bottom-Up DP)
# Time Complexity: O(m * n) - Two nested loops.
# Space Complexity: O(m * n) - For the 2D DP table.
def lcs_tabulation(text1: str, text2: str) -> int:
    """
    Computes the length of the Longest Common Subsequence (LCS)
    using iteration with tabulation (bottom-up DP).
    dp[i][j] stores the length of the LCS of text1[:i] and text2[:j].
    """
    len1, len2 = len(text1), len(text2)
    # dp table with (len1+1) rows and (len2+1) columns
    # dp[i][j] will store the LCS length for text1 up to index i-1 and text2 up to index j-1
    dp: List[List[int]] = [[0] * (len2 + 1) for _ in range(len1 + 1)]

    # Fill the DP table
    for i in range(1, len1 + 1):
        for j in range(1, len2 + 1):
            # If characters match, extend the LCS by 1 from the diagonal element
            if text1[i - 1] == text2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
            else:
                # If characters don't match, take the maximum from skipping a character
                # from text1 (dp[i-1][j]) or text2 (dp[i][j-1])
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[len1][len2]

# Space-optimized version for LCS (using O(min(m,n)) space)
# Time Complexity: O(m * n)
# Space Complexity: O(min(m, n)) - Uses two rows, where one row is min(len1, len2)
def lcs_space_optimized(text1: str, text2: str) -> int:
    """
    Computes the length of the Longest Common Subsequence (LCS)
    using space-optimized iteration (bottom-up DP).
    Reduces space complexity from O(m*n) to O(min(m,n)).
    """
    # Ensure text1 is the shorter string for space optimization
    if len(text1) < len(text2):
        text1, text2 = text2, text1

    len1, len2 = len(text1), len(text2) # len1 >= len2 now

    # dp array for current row, prev_dp for previous row
    # Each row corresponds to a column (len2)
    prev_dp: List[int] = [0] * (len2 + 1)
    curr_dp: List[int] = [0] * (len2 + 1)

    for i in range(1, len1 + 1): # Iterate through rows (length of text1)
        for j in range(1, len2 + 1): # Iterate through columns (length of text2)
            if text1[i - 1] == text2[j - 1]:
                # Diagonal element from prev_dp, plus 1
                curr_dp[j] = 1 + prev_dp[j - 1]
            else:
                # Max of element above (from prev_dp) or element to the left (from curr_dp)
                curr_dp[j] = max(prev_dp[j], curr_dp[j - 1])
        # After filling curr_dp for current 'i', it becomes prev_dp for the next 'i'
        prev_dp = list(curr_dp) # Make a copy

    return prev_dp[len2]


# --- Problem 4: Coin Change Problem (Minimum Coins) ---

# Problem Description:
# You are given an integer array `coins` representing coin denominations and an integer `amount`
# representing a total amount of money.
# Return the fewest number of coins that you need to make up that amount.
# If that amount cannot be made up by any combination of the coins, return -1.
# You may assume that you have an infinite number of each kind of coin.

# Approach 1: Recursive with Memoization (Top-Down DP)
# Time Complexity: O(amount * N) - Where N is number of coin denominations.
# Space Complexity: O(amount) - For the memoization table (dictionary) and recursion stack.
def coin_change_memoized(coins: List[int], amount: int) -> int:
    """
    Solves the Coin Change problem (minimum coins) using recursion with memoization.
    """
    memo: Dict[int, int] = {}

    def solve(target: int) -> int:
        # Base Cases:
        if target == 0:
            return 0  # 0 coins needed for amount 0
        if target < 0:
            return float('inf') # Cannot make a negative amount

        # Check memoization table
        if target in memo:
            return memo[target]

        min_coins = float('inf')
        for coin in coins:
            # Try using each coin
            res = solve(target - coin)
            if res != float('inf'): # If subproblem is solvable
                min_coins = min(min_coins, 1 + res) # 1 (for current coin) + min coins for remaining amount

        # Store and return the result
        memo[target] = min_coins
        return min_coins

    result = solve(amount)
    return result if result != float('inf') else -1

# Approach 2: Iterative with Tabulation (Bottom-Up DP)
# Time Complexity: O(amount * N) - Where N is number of coin denominations.
# Space Complexity: O(amount) - For the 1D DP table.
def coin_change_tabulation(coins: List[int], amount: int) -> int:
    """
    Solves the Coin Change problem (minimum coins) using iteration with tabulation.
    dp[i] represents the minimum number of coins to make up amount i.
    """
    # Initialize dp array. dp[0] = 0 (0 coins for amount 0).
    # All other values are float('inf') initially, as we haven't found a way to make them.
    dp: List[float] = [float('inf')] * (amount + 1)
    dp[0] = 0

    # Iterate through each possible amount from 1 to `amount`
    for i in range(1, amount + 1):
        # For each amount, consider all available coin denominations
        for coin in coins:
            # If the current coin can be used to form amount 'i'
            if i - coin >= 0:
                # Update dp[i] with the minimum of its current value
                # and (1 + dp[i - coin])
                # (1 for the current coin, plus the minimum coins needed for the remaining amount)
                dp[i] = min(dp[i], 1 + dp[i - coin])

    # If dp[amount] is still float('inf'), it means the amount cannot be made.
    return int(dp[amount]) if dp[amount] != float('inf') else -1

# Note: For Coin Change (minimum coins), the tabulation approach is already
# space-optimized to O(amount) because `dp[i]` only depends on `dp[i - coin_val]`,
# which are values already computed in the `dp` array. No separate space-optimized
# version is typically needed as `fibonacci_space_optimized` or `lcs_space_optimized`.